{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyN7VijP8LFDX1w5McGhG1Lx",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/LukeTheGeneWriter/gene_system_pHMMer/blob/main/gene_system_pHMMer.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Luke Williams -- 2/8/2025\n",
        "\n",
        "Welcome to the gene system pHMMer. This script works by walking through a gene file system and creating a profile hidden markov model of the sequences found in each gene folder.\n",
        "\n",
        "This allows follow-on analysis from the file system generated by AutoClipper, found at https://github.com/LukeTheGeneWriter/AutoClipper\n",
        "\n",
        "This script is pretty plug-and-play, but you do need to edit some path variables so that the program knows where the genetic file system is located, where it can write temporary files, and where it should output the final pHMM."
      ],
      "metadata": {
        "id": "mwwmwBNEUt4N"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Setup environment"
      ],
      "metadata": {
        "id": "nZwovTkPOhiY"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ulw-YPB7NrN4"
      },
      "outputs": [],
      "source": [
        "!pip install -q condacolab\n",
        "!pip install biopython\n",
        "import condacolab\n",
        "condacolab.install()\n",
        "!conda install -c bioconda pyhmmer\n",
        "!conda install bioconda::clustalw"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive', force_remount = True)\n",
        "%ls\n",
        "%cd drive\n",
        "%cd MyDrive\n",
        "%ls\n",
        "\n",
        "import os\n",
        "import pyhmmer\n",
        "import numpy as np\n",
        "from Bio import SeqIO\n",
        "from Bio import Seq\n",
        "from Bio import AlignIO\n",
        "from Bio.SeqRecord import SeqRecord"
      ],
      "metadata": {
        "id": "Lk002YU1Oqs9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "!!! You'll need to set these paths to connect this program to the data you want to process. Set the dest, interdest, and genesysdest variables."
      ],
      "metadata": {
        "id": "FzW45dtFUa2q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Set the destination folder for the profile hidden markov models\n",
        "dest = 'Gene_Sysyem_HMMs'\n",
        "if not os.path.exists(dest):\n",
        "  os.makedirs(dest)\n",
        "\n",
        "#Set the intermediate folder for gene multi-seuence alignments\n",
        "interdest = 'Genes_Aligned'\n",
        "if not os.path.exists(interdest):\n",
        "  os.makedirs(interdest)\n",
        "\n",
        "#Set the gene system folder\n",
        "genesysdest = 'Genomes_By_Gene'\n",
        "\n",
        "genes = os.listdir(genesysdest)\n",
        "print(genes)"
      ],
      "metadata": {
        "id": "VDAWR9TtOzDM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Below is a function to format the sequence records since Clustal Omega does not tolerate duplicate names"
      ],
      "metadata": {
        "id": "F3X8ZwmmQY8t"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def rename_dups(g):\n",
        "  finrecs = []\n",
        "  seen_names = []\n",
        "  seen_ids = []\n",
        "  for record in SeqIO.parse(os.path.join(interdest, g), 'fasta'):\n",
        "    #check if we'e seen it before\n",
        "    if record.name in seen_names or record.id in seen_ids:\n",
        "      #Check how many times it appears in the list\n",
        "      n = 0\n",
        "      for i in range(0, len(seen_names)):\n",
        "        if record.name == seen_names[i] or record.id == seen_ids[i]:\n",
        "          n += 1\n",
        "      #create a new record with \"__\" + str(n) appended to the rec id and name\n",
        "      newrec = SeqRecord(record.seq, id = record.id + \"__\" + str(n), name = record.name + \"__\" + str(n), description = record.description)\n",
        "      seen_names.append(record.name)\n",
        "      seen_ids.append(record.id)\n",
        "      finrecs.append(newrec)\n",
        "    else:\n",
        "      newrec = SeqRecord(record.seq, id = record.id + \"__0\", name = record.name + \"__0\", description = record.description)\n",
        "      finrecs.append(newrec)\n",
        "      seen_names.append(record.name)\n",
        "      seen_ids.append(record.id)\n",
        "  #print(seen_names)\n",
        "  #print(seen_ids)\n",
        "  return finrecs\n",
        ""
      ],
      "metadata": {
        "id": "YJTwzH1YQNK_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "This is the main"
      ],
      "metadata": {
        "id": "bZp1dt97RwVm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "for g in genes:\n",
        "  #Skip genes which are already computed\n",
        "  if os.path.exists(os.path.join(dest, g + \"_hmm\")):\n",
        "    print(\"Skipping \" + g)\n",
        "    continue\n",
        "\n",
        "  #Aggregate the fastas into one fasta\n",
        "  with open(os.path.join(interdest, g), 'w') as fyle:\n",
        "    for strain_file in os.listdir(os.path.join(genesysdest, g)):\n",
        "      strain_file_path = os.path.join(genesysest, g, strain_file)\n",
        "      for record in SeqIO.parse(strain_file_path, 'fasta'):\n",
        "        SeqIO.write(record, fyle, 'fasta')\n",
        "\n",
        "  #Set clustalw paths\n",
        "  clustal_in = os.path.join(interdest, g + \"_prepped\")\n",
        "  clustal_out = os.path.join(interdest, g + \"_alig\")\n",
        "\n",
        "  #Rename duplicates -- clustalw does not tolerate multiple seqs with the same name\n",
        "  non_dup_recs = rename_dups(g)\n",
        "\n",
        "  #Write one big fasta to put into clustal\n",
        "  with open(clustal_in, 'w') as c_in:\n",
        "    for r in non_dup_recs:\n",
        "      SeqIO.write(r, c_in, 'fasta')\n",
        "\n",
        "  #Run Clustal Omega to align the sequences\n",
        "  !clustalw -INFILE={clustal_in} -OUTFILE={clustal_out}\n",
        "\n",
        "  #Create pHMM from MSA\n",
        "  hmm_in = clustal_out\n",
        "  hmm_out = os.path.join(dest, g + \"_hmm\")\n",
        "  alphabet = pyhmmer.easel.Alphabet.dna()\n",
        "\n",
        "  with pyhmmer.easel.MSAFile(hmm_in, digital=True, alphabet=alphabet) as msa_file:\n",
        "    msa = msa_file.read()\n",
        "    interstring = g + \"_msa\"\n",
        "    msa.name = interstring.encode()\n",
        "    builder = pyhmmer.plan7.Builder(alphabet)\n",
        "    background = pyhmmer.plan7.Background(alphabet)\n",
        "    hmm, _, _ = builder.build_msa(msa, background)\n",
        "    hmm.consensus\n",
        "    #save\n",
        "    with open(hmm_out, 'wb') as hmmout_file:\n",
        "      hmm.write(hmmout_file)"
      ],
      "metadata": {
        "id": "nTlV0gB2RpIA"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}